File: regression/neural_network.R
# This script creates and evaluates a neural network in R for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering and parameter tuning.

# LOAD LIBRARIES ----------------------------------------------------------
# This section loads relevant libraries for this script.
library(readr)
library(data.table)
library(caret)

# IMPORT DATA -------------------------------------------------------------
# This section imports data.
# The data files are located in the 'R' branch file, NOT the 'classification' or 'regression' folders.
train = data.table(read_csv('train.csv'))
test = data.table(read_csv('test.csv'))

# CREATE MODEL ------------------------------------------------------------
# This section creates a neural network.

  # Set seed for reproducibility.
  set.seed(555)

  # Create neural network and include 5 fold cross validation.
  # The 'linout = T' argument is needed for regression.
  # This throws an error (not sure why), but seems to work fine in terms of prediction.
  nn = train(Sepal.Length ~ Sepal.Width + Petal.Length + Petal.Width, 
             train, 
             method = 'nnet', 
             linout = T,  
             trControl = trainControl(method = 'cv', 
                                      number = 5))
  
  # View model summary.
  nn
  
  # View a plot of hidden units, RMSE, and weight decay.
  plot(nn)
  
  # Generate training and test set predictions
  train$nn_pred = predict(nn, 
                          train)
  test$nn_pred = predict(nn, 
                         test)

# EVALUATE MODEL ----------------------------------------------------------
# Evalute 'caret' neural network by computing training, cross validation, and test set root mean square error (RMSE).
# Note that I do not include degrees of freedom in the training and test set calculations below.
# I'm not sure how 'caret' computes RMSE in the cross validation calculation.
  
  # Compute training set RMSE
  sqrt(sum((train$Sepal.Length - train$nn_pred) ^ 2) / nrow(train))
  
  # Compute cross validation set RMSE.
  mean(nn$resample$RMSE)
  
  # Compute test set RMSE.
  sqrt(sum((test$Sepal.Length - test$nn_pred) ^ 2) / nrow(test))

File: regression/neural_network.py
# This script creates and evaluates a neural network regression in Python for the 'iris' data.
# This iteration is only to get the model up and running, so there is no feature engineering or parameter tuning.

################
# IMPORT MODULES
################
# This section imports modules needed for this script.
import os
import pandas as pd
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

#############
# IMPORT DATA
#############
# This section imports data.
# The data files are located in the 'Python' branch file, NOT the 'classification' or 'regression' folders.
# The commands below assume the working directory is set to the 'classification' folder.
os.chdir('..')
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

##############
# CREATE MODEL
##############
# This section creates a neural network regression model.

# Create and fit neural network regression model.
nn_reg = MLPRegressor(max_iter = 500,
                      random_state = 555).fit(train[['Sepal.Width', 'Petal.Length', 'Petal.Width']],
                                              train['Sepal.Length'])

# View current loss.
nn_reg.loss_

# View the number of iterations carried out.
nn_reg.n_iter_

# View the number of layers.
nn_reg.n_layers_

# View name of output activation function (check documentation).
nn_reg.out_activation_

# Conduct 3 fold cross validation.
cv = cross_val_score(nn_reg,
                     train[['Sepal.Width', 'Petal.Length', 'Petal.Width']],
                     train['Sepal.Length'],
                     cv = 3,
                     scoring = 'neg_mean_squared_error')

# Generate training and test set predictions.
train['nn_reg_pred'] = nn_reg.predict(train[['Sepal.Width', 'Petal.Length', 'Petal.Width']])
test['nn_reg_pred'] = nn_reg.predict(test[['Sepal.Width', 'Petal.Length', 'Petal.Width']])

################
# EVALUATE MODEL
################
# This section computes the mean squared error of the model on the training, cross validation, and test sets.

# Compute mean squared error on training set.
mean_squared_error(train['Sepal.Length'],
                   train['nn_reg_pred'])

# Compute mean squared error on cross validation set.
cv.mean() * -1

# Compute mean squared error on test set.
mean_squared_error(test['Sepal.Length'],
                   test['nn_reg_pred'])
